change type of numerical architecture symbols to integer

kroening opened the pull request:
Following a suggestion in ##5878, this changes the type of the numerical __CPROVER_architecture_ symbols from `int` (which is itself architecture-dependent) to integer (which does not depend on the architecture).

Closes #5878

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
change type of numerical architecture symbols to integer

Following a suggestion in ##5878, this changes the type of the numerical
__CPROVER_architecture_ symbols from int (which is itself
architecture-dependent) to integer (which does not depend on the
architecture).

Closes #5878

tautschnig commented:
Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).

kroening commented:
> Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).

No need -- this is compatible both ways (old versions can read new binaries, and old binaries can be read by new versions).

tautschnig commented:
> > Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).
> 
> No need -- this is compatible both ways (old versions can read new binaries, and old binaries can be read by new versions).

Linking won't work for the architecture symbols now have conflicting types. We might gloss over that and would then just need to recompile the goto binaries in our regression tests.

Further fixes that seem to be necessary:
1. Intervals don't know about `integer_typet`, which can be fixed as follows:
````patch
--- a/src/util/interval.cpp
+++ b/src/util/interval.cpp
@@ -1094,7 +1094,7 @@ bool constant_interval_exprt::is_numeric(

 bool constant_interval_exprt::is_int(const typet &type)
 {
-  return (is_signed(type) || is_unsigned(type));
+  return is_signed(type) || is_unsigned(type) || type.id() == ID_integer;
 }

 bool constant_interval_exprt::is_float(const typet &src)
````
2. The C++ front-end rejects the type conversion in the assignment of the architecture string.

kroening committed:
C++: __CPROVER_integer and __CPROVER_rational

This extends the C++ front-end to recognize __CPROVER_integer and
__CPROVER_rational, in the same way as done in the C frontend.

kroening committed:
constant_interval_exprt now recognizes integer_typet

This enables using unbounded integers instead of bitvectors for
architecture-independent constants.

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -113,7 +113,8 @@ static std::string architecture_string(const std::string &value, const char *s)
 static std::string architecture_string(T value, const char *s)
 {
-  return std::string("const int " CPROVER_PREFIX "architecture_") +
+  return std::string("const __CPROVER_integer " CPROVER_PREFIX
```
I think this should also use `CPROVER_PREFIX` in the first case of "\_\_CPROVER_" (actually the failing cpplint job says the same :-) )
